### ISSUE TITLE
Power Supply determine all needed commands supported in DD&simics

### ISSUE THREAD

**Issue by derekhoward55** [ibm-openbmc/dev#1424]
Determine which cmds/registers we need to use. 
Ensure device drivers are supported for them, ask for support for ones missing.
Send complete list to simics to ensure simics supports whether it is something we'd like, or something the dd is doing under the covers.

Rainier 2300w ps have ccin 51E9, which have the 2-byte version reads.
Early bringup will need support for PDQ ps, ccin 51DE, which have 1-byte version reads like 1400W ps with ccin 2B1E.  Ensure no other side-effects of using PDQ ps with the DD.

**Comment by bjwyman** [comment]
Reviewed commands we plan or do not plan to use with @derekhoward55.

Device driver may require some commands in order to load correctly, built on top of pmbus-core code.

SIMICS may have to return sensible values for some commands, and default to no fault bits on for others.

|Command|Length|PMBus Name|Notes|
|-------|------|----------|-----|
| 01h | 0 - 1 | OPERATION | No plans to require device driver or simulator support. *Verify NOT used by Witherspoon.* |
| 02h | 0 - 1 | ON_OFF_CONFIG | Desirable to have as accessible to at least write via the device driver. We would want to ensure that this is written to power on the supplies via the GPIO enable, not requiring the 01h command, when the PSU monitor code starts up (initialization). Simulator should at least support writing a value. |
| 03h | 0 | CLEAR_FAULTS | A write clears faults. Understood that phosphor-hwmon *should* be essentially doing this once a second, but desirable to have for initialization and clearing of latched AC loss faults. SIMICS should already support, due to pmbus-core code used by ibm-cffps device driver. The ibm-cffps device driver should allow for write. |
| 7Ch | 1 | STATUS_INPUT | Assuming SIMICS already needs to support this, due to pmbus-core code. Only potentially read for error/debug data. |
| 80h | 1 | STATUS_MFR_SPECIFIC | Read for various fault debug data. Check on device driver bit6 mapping to VOUT fault. |
| 81h | 1 | STATUS_FANS_1_2 | Likely included as pmbus-core support. Not required by PSU monitor code, no plans to read/check. |
| 88h | 2 | READ_VIN | Needed. Simulator should return a good/sensible value based on the power supply and system type. |
| 97h | 2 | READ_PIN | Needed. Simulator should return a good/sensible value based on the power supply and system type. |
| 9Bh | 13 | MFR_REVISION | Needed. Part number for VPD (inventory). |
| 9Ch | 7 | MFR_LOCATION | Needed. "VPD Header". Not clear on if good/sensible value required. |
| 9Eh | 7 | MFR_SERIAL | Needed. Serial number for VPD (inventory). |
| A7h | 2 | MFR_POUT_MAX | Needed. Device driver should support ability to read. Simulator should return a good/sensible value based on power supply and/or system type. |
| BDh | 2 | USER_DATA_13 | CCIN (VPD, inventory). Device driver uses to determine firmware version format. Simulator should set to a good/sensible avlue based on power supply and/or system type. |
| BEh | 1 - 2 | USER_DATA_14 | Should **NOT** require simulator or device driver support. Request power supply specification(s) map PSKill fault to 80h. |
| D6h | 6 - 101 | MFR_SPECIFIC_D6, INPUT_HISTORY | Needed. Input history table. Telemetry data.|
| D7h | 0 - 1 | MFR_SPECIFIC_D7, INPUT_HIST_SYNC | *Not clear if needed or if FFS/SYNC GPIO usage sufficient.* |
| DAh | 1 | MFR_SPECIFIC_DA | Simulator needs to support writing. Device driver already uses for LED control. |
| DEh | ? | MFR_SPECIFIC_DE | **NOT** needed by PSU monitor, but **NEEDED** for PSU firmware update. Device driver should support read. Simulator should support reading a good/sensible value. |
| FAh | 1 - 2 | FW_VERSION | Firmware revision. Device driver bases 1 or 2 byte read using BDh value. SIMICS should support one or two byte reads. |
| FBh | 1 - 2 | FW_VERSION | Firmware revision. Device driver bases 1 or 2 byte read using BDh value. SIMICS should support one or two byte reads. |
| FCh | 1 - 2 | FW_VERSION | Firmware revision. Device driver bases 1 or 2 byte read using BDh value. SIMICS should support one or two byte reads. |
| FDh | 1 | FW_VERSION, Reserved? | Firmware revision or reserved, depending on power supply type. Device driver bases 1 or 2 byte read using BDh value. SIMICS should support one or two byte reads. |

**Comment by bjwyman** [comment]
Still working with power supply engineer and simulator developers on BEh and D6h. Hoping to not require the BEh command in the simulator or the device driver. Ability to read D6h data that is not a bad format requested, with ability to modify it in the simulator.

**Comment by eddiejames** [comment]
I've sent a series upstream for this: https://patchwork.kernel.org/project/linux-hwmon/list/?series=219931

I did not include a debugfs file to manually write to clear_faults. The pmbus core does that automatically when it reads it's set of registers.

**Comment by derekhoward55** [comment]
I thought Brandon described only the app that reads the input power goes down the path where the clear_faults get sent, but not the psu-monitor app (the psu-monitor code that does the reads goes down a different path in which the clear_faults is not sent every second)?  If that is in fact the case, and we can't do the clear faults from the psu-monitor app, then we'll have a dependency on the other app (we'll need to make sure it is always running when the psu-monitor app is, can't be held off, etc).

**Comment by eddiejames** [comment]
Not sure what those different apps do, but if you read any of the normal pmbus properties it will clear the faults, so you could always just do that.

**Comment by bjwyman** [comment]
I thought we looked back over the device driver code to determine how it was that CLEAR_FAULTS got in there and it had something to do with the path to the files. The hwmon process looks at those files like in1, but the PSU application looks at some other debug files, due to how most of the other stuff we wanted, not hwmon related, had information we needed. Somehow that debug file path did not involve the path that called CLEAR_FAULTS.

**Comment by bjwyman** [comment]
```
.../sys/kernel/debug/pmbus/hwmon3# ls
cffps1         status0_cml    status0_input  status0_mfr    status0_temp
status0        status0_fan12  status0_iout   status0_other  status0_vout
.../sys/kernel/debug/pmbus/hwmon3# grep . *
status0:0x0840
status0_cml:0x00
status0_fan12:0x00
status0_input:0x00
status0_iout:0x00
status0_mfr:0x00
status0_other:0x00
status0_temp:0x00
status0_vout:0x00
```
Doing something like that apparently goes through the debugfs functions.

```
static ssize_t pmbus_show_sensor(struct device *dev,
                                 struct device_attribute *devattr, char *buf)
{
        struct pmbus_data *data = pmbus_update_device(dev);
        struct pmbus_sensor *sensor = to_pmbus_sensor(devattr);

        if (sensor->data < 0)
                return sensor->data;

        return snprintf(buf, PAGE_SIZE, "%ld\n", pmbus_reg2data(data, sensor));
}
```
That code path gets involved when you read these:
```
.../sys/bus/i2c/devices/3-0068/hwmon/hwmon3# ls
curr1_alarm   fan1_alarm    fan2_alarm    in1_input     in2_label     power1_input  temp1_input   temp3_input
curr1_input   fan1_fault    fan2_fault    in1_label     name          power1_label  temp2_alarm   uevent
curr1_label   fan1_input    fan2_input    in2_alarm     of_node       subsystem     temp2_input
device        fan1_target   in1_alarm     in2_input     power1_alarm  temp1_alarm   temp3_alarm
```

**Comment by bjwyman** [comment]
https://patchwork.kernel.org/patch/11304393/
Commenting after series merged.

Guenter seemed to have some concerns about 02h allowing for a write. In most cases, I would say that we plan NOT to write that. The concern here would be some PSU firmware blunder defaulting it contrary to what was expected, or some developer/engineer inadvertently changing it via i2ctools and not setting it back to the default.

I had not noticed that 9Ch was missing from the DD, but see it added now. A7h was also added. Both of those are just in the read path, which makes sense, we should not be writing to those.

I am not certain what the 80h bit 6 part was about, no changes noticed there, which is probably okay.

No updates for BEh, D7h, or DEh. The first two seem right, but not sure if DEh missing is bad, as far as firmware update support goes.

**Comment by bjwyman** [comment]
DEh was already in the device driver. :-)
https://github.com/openbmc/linux/blob/f7afc39e5727f68242a0d466e51a56d7d9c59337/drivers/hwmon/pmbus/ibm-cffps.c#L31

**Comment by bjwyman** [comment]
BEh PSKill fault still planned on moving to 80h.

**Comment by bjwyman** [comment]
I think I saw an indication of 80h now having the bit that was (also) in BEh.
Still waiting on replies for CLEAR_FAULT.

**Comment by bjwyman** [comment]
If any of the "hwmon" type file paths are read, that should force the CLEAR_FAULT command to be sent. The plan for now will be to read something like `in1_alarm` or `in1_input` to force the CLEAR_FAULT if we want to make sure that command is sent. Probably with a comment about why that is how we choose to clear the faults instead of having a specific debug file for CLEAR_FAULT itself, since that was not put into the device driver.

**Comment by bjwyman** [comment]
If the device driver is updated to recognize the Rainier CCIN, and configure things up for 2 page support, and 3 x 2-byte reads for firmware version, that should work. 

If the PDQ CCIN is read, and no match is found, it looks like the device driver should just default to 1 page support, and four 1-byte reads for firmware version. That at first glance seems like it should work.

**Comment by bjwyman** [comment]
https://patchwork.kernel.org/patch/11432399/

**Comment by bjwyman** [comment]
Not merged into Linux kernel fork openbmc is using, yet.

**Comment by bjwyman** [comment]
Okay, I think recognizing the new CCIN was the last of this we needed.
https://lists.ozlabs.org/pipermail/openbmc/2020-April/021118.html

https://github.com/openbmc/linux/commit/962db044aa3f3f51405fb6a956bc4d5ac77463ba#diff-225e4f85c83ed9ffae2e26e5e2e9f0c5
